### PR TITLE
Fixed issue with collapsable ribbon and quick acces toolbar

### PR DESCRIPTION
**Pull request by APEbbers**
Hi,

I went a bit futher with the ribbon. I noticed that when you collapse the ribbon, the area didn't shrink with it. Also when I opened FreeCAD on my laptop screen, the quickaccess toolbar dissapeard.

I totally understand if I overstept any boundary, so feel free to ignore all below. I can imagen that you have other plans with this workbench.

The biggest change is that I removed the customized qtpy folder.
After a few attempts, I was stuck and lost a bit of track with all the changes that I had done. I was curious to see what happend if i use the original qtpy package. As it turns out, the ribbon collapsed as it should. So I went a bit furtner with the changes below.

I've added the following:
-  the width of the quick access toolbar is now based on the number of buttons.
- the height of the quick access toolbar and the application button are tweaked a little bit.
- I Moved the "geticonsize" to the class level. This way you can use it with all defs in the class
- the tabs are now sorted based on the order of the workbenches as defined in preferences->workbenches
- I've added you customized stylesheets and have the qtpy defaults overridden with it.
-  I've added qtpy and pyqtribbon as a dependeny in package.xml. All customizations are now done in FCBinding.py. 

kind regards

Paul

### CODE REVIEW

**Comment by geolta** [comment]
Hi, these look like nice quality of life improvements :)
Unfortunately I'm currently in vacation without my laptop and therefore not able to properly review your changes. But I promise you to look at them when I'm back (in ~3 weeks).
Keep your improvements flowing, they are pushing this repository from a testing state to something usable, maybe we can then even add it to the Addon Manager :)
Thank you again!

**Comment by APEbbers** [comment]
Hi,

a short update about my progress. I've created a settings dialog which you can open via a button under the application menu.
With this dialog, you can change the ribbon design. you can change the ribbon buttons per workbench and then per toolbar.
To achieve this, I've added two levels above "Toolbars" -> "workbenches" & "workbench name".
I now people prefer a drag/drop function, but I figured that with this dialog in place, we have more time to look into that.

When you open the settings dialog for the first time, all workbenches will be loaded. I noticed that when you don't activate a workbench, you do not get its toolbars and commands properly. So it takes a little bit of time to open at the moment.

Overall I changed/added the following:
- Added a helpbutton with the link to the help as set in the FreeCAD preferences
- Added a settings button under the application button to open the settings dialog
- Added a separate file called "Parameters_Ribbon.py". I used this to centralize all settings that we can use in a preference page.
- Added a file with standard functions. (Is a collection of functions that I created for my workbenches)
- Added a settings dialog with the following functions:
   - Select quick access commands
   - In/Exclude workbenches
   - In/Exclude toolbars
   - Set the Icon size for commands per workbench, per toolbar
   - Buttons to change the order of commands
   - All Icons are now from FreeCAD itself including dropdown buttons (I took the first command in the dropdown)
   - Restore function.
   - Reset function. (a proper default layout has to be made yet)
   - Settings tab with the folowing settings:
       - Enable/Disable the backup function
       - Set the backup location
       - Enable or disable AutoHide
       - Change the size of the small, medium and large buttons
       - Load a stylesheet 

To generate a default layout for the reset button I'm thinking to create a separate script, that set every first command in a toolbar to large. What is your opinion about this? Or we make a default for a few internal workbenches and the rest is up to the users?

I didn't started with a preference page yet. I've made two workbenches so far and honestly I don't like the preference page that much. It is limited in what you can do with it.

kind regards

**Comment by APEbbers** [comment]
I've added a script to generate a json file. It is standalone, so you can run it directly in FreeCAD.
The script set every first command of a toolbar to large. 
I believe it helps to create a default for the ribbon.

The next thing I'd like to do is add the possibility to combine toolbars to one big toolbar.
I think if we have that, we have a good base to release the ribbon as an add-on. And develop further from there.

**Comment by APEbbers** [comment]
You can now sort the panels, create custom panels and custom toolbars are now included

**Comment by geolta** [comment]
Hi, thank you again for your work and sorry for my late reply :)

The problem is, that this PR does a lot of things at once and has 15k lines of code added (although most of them are the JSON file and the QtDesigner UI files). This is not reviewable for me. Is it feasible for you to split this into multiple smaller PRs? I totally understand if this is too much work for you. Then I would limit my review to thorough testing and a few important code snippets :)

Best regards!

**Comment by APEbbers** [comment]
Hi,

 

I don’t think it is possible to split up the pull request. But maybe you can create a second branch? I then create a pull request to that one, so you can review it without changing the main branch?

 

Best regards!

**Comment by APEbbers** [comment]
Hi,

Did you had a look at my updates? Fixed a few issues. The way I loaded the ribbon, caused the freecad preferences to crash.
So I went back to you way of implementing the ribbon. I added custom functions to show/hide and autohide the ribbon.

I also added the menus from the standard menubar to the application button. 
image

With the "alt" key you can show/hide the standard menubar.

what do you think of my approach so far? I think most issues are handled without changes needed for all workbenches.
On the forum they are going into another direction, but I quite like the use of the ribbon, so maybe we can release it as an add-on?

kind regards

**Comment by geolta** [comment]
> I don’t think it is possible to split up the pull request. But maybe you can create a second branch? I then create a pull request to that one, so you can review it without changing the main branch?

I'm not sure, whether I fully understand your suggestion. What would be the advantage of a second branch?


Very nice that the menu bar is now squashed into the FreeCAD icon!

I think, it definitely makes sense to publish this as an addon for the Addon Manager. If feasible, it would be great if we could manage to do this before the 1.0 release (sorry again for my late replies).
Do you have experience in bringing stuff to the Addon Manager?

I will have a detailed look at your changes tomorrow.
Feature wise I think they are quite neat :)

**Comment by APEbbers** [comment]
You right, you can also test my fork directly.

I've created 2 addons previously, so releasing it to the addon manager is no problem. I'll create a wiki page first, so that there is a manual in place? 

btw:
I added a dependency to the package.xml: keyboard. This is needed for detecting the alt key press.
I've also implemented a scroll function for when the ribbon has more panels than that fit on your screen

**Comment by geolta** [comment]
Hi, I have tested most of your changes. Here are my remarks :)

- Installation: The Addon Manager cannot install the `pyqtribbon` and `keyboard` libraries, at leat not on the AppImage for linux (but it can install `pyqt` for whatever reason). I had to manually copy the libraries into the project folder. If I remember correctly, this also was the reason to include `pyqtribbon` directly.
- The keyboard library seem not to work on Wayland under linux (tested on Gnome 46). It says, I have to be the admin to execute `on_press_key`.
- The help buttons in Design dialog and in the ribbon bar seem not to work. This might also be some permission issue.
- It took me some time to figure out what the pin button of the ribbon does – adding a tooltip for this would be good
- Making every first button of a toolbar large, seems not to be the best idea to me. This leads to many strange designs like the point in Sketcher being large. I understand, that this is just a starting point, but I think, keeping everything small at the beginning would be better. But of course, in the end, it's a matter of taste.
- Adding a warning like "may take some time" to the "Open Design" button might be good
- Add hint to Design dialog that a restart is required for the changes to take effect.
- I don't know from my head how to get icons from FreeCAD instead of having them "hard coded", but I think that this might be better to also be able to automatically apply icon themes, etc.
- Is there a reason you have changed every name, author, maintainer **everywhere** across the entire project to you? I understand this for modules written by you, but we are still relying on work from HakanSeven and me.
- The README should get an update after all of these changes.

Best regards :)

**Comment by APEbbers** [comment]
Hi thanks for testing! 

I've written my response per comment below:

> "Installation: The Addon Manager cannot install the pyqtribbon and keyboard libraries, at least not on the AppImage for linux (but it can install pyqt for whatever reason). I had to manually copy the libraries into the project folder. If I remember correctly, this also was the reason to include pyqtribbon directly."

I wasn't aware of the issues with AppImage. I'll look into that. Maybe also a right issue? We can always include the pyqtribbon again


> "The keyboard library seem not to work on Wayland under linux (tested on Gnome 46). It says, I have to be the admin to execute on_press_key.
The help buttons in Design dialog and in the ribbon bar seem not to work. This might also be some permission issue."

I've used packages that are standard included in python, but I can look into different packages. I'm working mostly on windows, but I have a chromebox with debian on it.


> "It took me some time to figure out what the pin button of the ribbon does – adding a tooltip for this would be good"

agreed, I'll add a tooltip for it

> "Making every first button of a toolbar large, seems not to be the best idea to me. This leads to many strange designs like the point in Sketcher being large. I understand, that this is just a starting point, but I think, keeping everything small at the beginning would be better. But of course, in the end, it's a matter of taste."

The script allows to add already customized workbenches. I'm thinking to add customization for internal workbenches as standard like the sketcher WB. For every external workbench we have than the first icon of every toolbar set to large.
After that, the user can customize the script as desired. (or add extra scripts to the folder)

> "Adding a warning like "may take some time" to the "Open Design" button might be good"

agreed. I'll add a warning

> "Add hint to Design dialog that a restart is required for the changes to take effect."

agreed. I add the same or similar dialog as used with the add-on manager

> "I don't know from my head how to get icons from FreeCAD instead of having them "hard coded", but I think that this might be better to also be able to automatically apply icon themes, etc."

Actually, all icons are now retrieved from FreeCAD. But they are still written to the json file. This way we can add custom icon functionality easily at a later point. The way it works is that the design dialog retrieves the icons from freecad and writes the name to the json file. FCBinding.py, first retrieves the icon from FreeCAD. Then tries to get the icon from the Json file. If this succeeds, that icon is used. If it fails, the icon from freecad is used.

 

> "Is there a reason you have changed every name, author, maintainer everywhere across the entire project to you? I understand this for modules written by you, but we are still relying on work from HakanSeven and me."

i went a bit over board there. i noticed that the license in the files didn't match the MIT license, so I changed that and copied that to all files. I'll change the text to include both  your names as well? 

> "The README should get an update after all of these changes."

Agreed. This can be done when the wiki will be written.


kind regards

**Comment by geolta** [comment]
Hi Paul,

> I wasn't aware of the issues with AppImage. I'll look into that. Maybe also a right issue? We can always include the pyqtribbon again

I think, this is not a rights issue, but rather a matter of which python executable is used and since the AppImage bundles its own python it doesn't access the packages installed system wide.

Maybe we can use git submodules to include pyqtribbon – but to be honest, I have never used them :)
  
> I've used packages that are standard included in python, but I can look into different packages. I'm working mostly on windows, but I have a chromebox with debian on it.

Maybe just add an if clause to the keyboard stuff to check that we're not running linux?

> The script allows to add already customized workbenches. I'm thinking to add customization for internal workbenches as standard like the sketcher WB. For every external workbench we have than the first icon of every toolbar set to large. After that, the user can customize the script as desired. (or add extra scripts to the folder)

You have convinced me :)

> > "I don't know from my head how to get icons from FreeCAD instead of having them "hard coded", but I think that this might be better to also be able to automatically apply icon themes, etc."
> 
> Actually, all icons are now retrieved from FreeCAD. But they are still written to the json file. This way we can add custom icon functionality easily at a later point. The way it works is that the design dialog retrieves the icons from freecad and writes the name to the json file. FCBinding.py, first retrieves the icon from FreeCAD. Then tries to get the icon from the Json file. If this succeeds, that icon is used. If it fails, the icon from freecad is used.

Ah, I meant the FreeCAD logo and things like the help button. Sorry for being imprecise.

> > "Is there a reason you have changed every name, author, maintainer everywhere across the entire project to you? I understand this for modules written by you, but we are still relying on work from HakanSeven and me."
> 
> i went a bit over board there. i noticed that the license in the files didn't match the MIT license, so I changed that and copied that to all files. I'll change the text to include both your names as well?

I have checked the license of HakanSevens repo again (https://github.com/HakanSeven12/Modern-UI?tab=GPL-3.0-1-ov-file#readme), it is a GPL 3.0 license so we're forced to use this as well. Maybe just use our three names? 

Best Regards! :)

**Comment by APEbbers** [comment]
Hi,

I fixed most of the issues:
- I added the pyqtribbon package as a backup. If this package is not installed, a local version will be used.
- I added a restart dialog to the design and preference dialogs. (freecad should now restart when clicked "restart")
- The keyboard library is now only imported when on windows or mac. (For linux I didn't find an alternative yet) 
   - For linux I used the Qt version for now. Only you have to click on the ribbon first to activate the function which is not ideal
- I added the tooltip to the pin button
- I hadn't problems with opening the help pages on linux. However:
   - I replaced the help button on the ribbon, with the one from FreeCAD itself
   - In the design dialog, the icon from the FreeCAD help is used.
   - I added an "About FreeCAD Ribbon" to the application menu->Customize
   - Both the help button in the design dialog and the about button refer to the repositories from which the ribbon is installed.
   - The wikipage on my repository was disabled. This is enabled now.
- I updated the license in both the files as in the repository 
 
I'll start with a draft design with the most important wb's (Part/PartDesign, TechDraw, Surface, BIM, FEM and CAM.
After that I'll update the default json file and the scripts.
Followed with the Wiki page and the readme

kind regards

File: README.md [hunk 1]
```diff
@@ -1,23 +1,43 @@
...
+* Much more...
+
+See the wiki page for more details and a manual on how to customize the Ribbon to your liking.
 
-This is not finished or intended for production use, but rather to look at where the limitations of this approach, implementing it as an external Python addon, lie.
 
 ## Installation
 There are two options for installing this addon:
 ### Custom Repository for Addon Manager
-Go to `Edit/Preferences/Addon-Manager` and add the custom repository `https://github.com/geolta/FreeCAD-Ribbon.git` with the branch `main`. Now you can go to the Addon Manager (`Tools/Addon-Manager`) and install "FreeCAD Ribbon" (maybe you have to update your local cache first). Now restart FreeCAD and you will see a ribbon interface :)
```
Comment (geolta) [inline README.md:12]: Please revert the change in the URL

File: README.md [hunk 2]
```diff
@@ -1,23 +1,43 @@
...
 
-A draft to implement a proper Ribbon UI for FreeCAD, based on the work of [HakanSeven](https://github.com/HakanSeven12/Modern-UI) and the [PyQtRibbon library](https://github.com/haiiliin/pyqtribbon) that got slight modifications for this purpose.
+An Ribbon UI for FreeCAD, based on the PyQtRibbon library (https://github.com/haiiliin/pyqtribbon).  
+This ribbon is based on the work of HakanSeven (https://github.com/HakanSeven12/Modern-UI) for the Modern-UI workbench.  
+Current developers are:
+* Paul Ebbers (https://github.com/APEbbers/FreeCAD-Ribbon)
+* Geolta (https://github.com/geolta/FreeCAD-Ribbon)
 
-![Screenshots of FreeCAD with the Ribbon UI](Screenshot.png)
+![](https://github.com/APEbbers/FreeCAD-Ribbon/blob/main/Resources/Images/Screenshot.png)
```
Comment (geolta) [inline README.md:6]: Please revert the change in the URL of the image

File: README.md [hunk 3]
```diff
@@ -1,23 +1,43 @@
...
+Go to `Edit/Preferences/Addon-Manager` and add the custom repository `https://github.com/APEbbers/FreeCAD-Ribbon.git` with the branch `main`. Now you can go to the Addon Manager (`Tools/Addon-Manager`) and install "FreeCAD Ribbon" (maybe you have to update your local cache first). Now restart FreeCAD and you will see a ribbon interface :)
 ### Manual Installation
 Download this repository, extract the folder and copy it to the `Mod` folder of FreeCAD, detailed information can be found at the [FreeCAD Wiki](https://wiki.freecad.org/Installing_more_workbenches). Now restart FreeCAD and you will see a ribbon interface :)
 
 ## Uninstallation
 1. Remove the folder of this in the `Mod` folder of your FreeCAD installation
 1. Restart FreeCAD.
 1. When you restarted you don't see any toolbar.
-1. Create a macro.
+1. Download the file
```
Comment (geolta) [inline README.md]: download which file?

File: Resources/Images/Screenshot.svg
Comment (geolta) [inline Resources/Images/Screenshot.svg:1]: Well, this image looks broken :D

File: package.xml [hunk 4]
```diff
@@ -5,27 +5,29 @@
...
-  <maintainer>geolta</maintainer>
+  <maintainer>Paul Ebbers, Geolta</maintainer>
 
-  <license file="LICENSE">GPL v3.0</license>
+  <license file="LICENSE">GPL-3.0 license</license>
 
   <pythonmin>3.8.0</pythonmin>
 
-  <url type="website">https://github.com/geolta/FreeCAD-Ribbon/wiki</url>
+  <url type="website">https://github.com/APEbbers/FreeCAD-Ribbon/wiki</url>
```
Comment (geolta) [inline package.xml:18]: Please keep the URL

File: package.xml [hunk 5]
```diff
@@ -5,27 +5,29 @@
...
 
-  <url branch="main" type="repository">https://github.com/geolta/FreeCAD-Ribbon</url>
+  <url branch="main" type="repository">https://github.com/APEbbers/FreeCAD-Ribbon</url>
```
Comment (geolta) [inline package.xml:20]: Here as well

File: package.xml [hunk 6]
```diff
@@ -5,27 +5,29 @@
...
 
-  <url type="readme">https://github.com/geolta/FreeCAD-Ribbon/blob/main/README.md</url>
-
-  <icon></icon>
+  <url type="readme">https://github.com/APEbbers/FreeCAD-Ribbon/blob/main/README.md</url>
```
Comment (geolta) [inline package.xml:22]: and here

**Review by geolta — COMMENTED** [review]
Hi, very nice!
As far as I can see, there are only a couple of very small issues :)
If added them as reviews to the code.
Best regards!

File: .vscode/settings.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+  "[python]": {
+    "editor.defaultFormatter": "ms-python.black-formatter"
+  },
+  "python.formatting.provider": "none",
+  "editor.fontLigatures": false,
+  "workbench.colorCustomizations": {},
+  "autopep8.interpreter": ["D:\\Python\\Python3.10\\python.exe"],
```
Comment (geolta) [inline .vscode/settings.json]: This explicit path is not good, for other people (on other OS) this will be different. Please remove it and set this path as a user setting for you.

File: RibbonStructure.json.bak
Comment (geolta) [inline RibbonStructure.json.bak:1]: Does it make sense to add the backup file to git?

File: Scripts/UpdateCommandNames.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            if ToolBar != "order":
+                for Command in Dict_RibbonCommandPanel["workbenches"][WorkBench][
+                    "toolbars"
+                ][ToolBar]["commands"]:
+                    for key, value in Dict_Commands.items():
+                        if Command == key:
+                            if value[2] != "" and value[2] != "...":
+                                Dict_RibbonCommandPanel["workbenches"][WorkBench][
+                                    "toolbars"
+                                ][ToolBar]["commands"][Command]["text"] = value[2]
```
Comment (geolta) [inline Scripts/UpdateCommandNames.py:133]: This function has four nested for-loops plus three nested if clauses. That is quite a lot :D
Please break this into smaller functions or at least add a few comments.

File: Standard_Functions_RIbbon.py [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+
+def GetLetterFromNumber(number: int, UCase: bool = True):
+    # from openpyxl.utils import get_column_letter
+
+    # Letter = get_column_letter(number)
+
+    # # If UCase is true, convert to upper case
+    # if UCase is True:
+    #     Letter = Letter.upper()
```
Comment (geolta) [inline Standard_Functions_RIbbon.py]: Why is this as a comment? If it is not needed, please remove it or leave a comment, why it is there as a comment.

File: Standard_Functions_RIbbon.py [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    while number > 0:
+        number, r = divmod(number - 1, 26)
+        Letter = chr(r + ord("A")) + Letter
+    return Letter
+
+
+def GetNumberFromLetter(Letter):
+    # from openpyxl.utils import column_index_from_string
+
+    # Number = column_index_from_string(Letter)
```
Comment (geolta) [inline Standard_Functions_RIbbon.py]: here as well

**Review by geolta — COMMENTED** [review]
I have now also skimmed the code and found a few minor issues.

File: .vscode/settings.json (hunk 7 above)
Comment (APEbbers) [reply .vscode/settings.json]: I've removed the folder and updated .gitignore. It shouldn;t be in git, but just local for me

File: RibbonStructure.json.bak
Comment (APEbbers) [reply RibbonStructure.json.bak:1]: indeed, old left over. I've removed it

File: Scripts/UpdateCommandNames.py (hunk 8 above)
Comment (APEbbers) [reply Scripts/UpdateCommandNames.py:133]: I've added some comments to clarify

File: Standard_Functions_RIbbon.py (hunk 9 above)
Comment (APEbbers) [reply Standard_Functions_RIbbon.py]: I've a standard functions repo where this is a copy of. This is probably a left over from my other wb's. it is removed.

File: Standard_Functions_RIbbon.py (hunk 10 above)
Comment (APEbbers) [reply Standard_Functions_RIbbon.py]: same as mentioned above

File: README.md (hunk 1 above)
Comment (APEbbers) [reply README.md:12]: Can you change this after you merged all the changes to your repo? I prefer to have not to have links to another repo in my own repro

File: README.md (hunk 2 above)
Comment (APEbbers) [reply README.md:6]: Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

File: package.xml (hunk 4 above)
Comment (APEbbers) [reply package.xml:18]: Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

File: package.xml (hunk 5 above)
Comment (APEbbers) [reply package.xml:20]: Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

File: package.xml (hunk 6 above)
Comment (APEbbers) [reply package.xml:22]: Can you change this when you merged all the changes to your repo? I prefer to have not to have links to another repo when I'm using my own repo for install

File: README.md (hunk 3 above)
Comment (APEbbers) [reply README.md]: Changed the text to: "Create a new macro (Macro menu -> macros...-> create macro)"

File: Resources/Images/Screenshot.svg
Comment (APEbbers) [reply Resources/Images/Screenshot.svg:1]: I'm no longer using, so I've removed it.

**Comment by APEbbers** [comment]
Hi,

I've finished the Wiki page as well.  I believe the addon is now ready to be added to the addon manager.
at the bottom of [https://wiki.freecad.org/Add_Workbench_to_Addon_Manager](url) you can find how to add this to the add-on manager.

Regarding the links in the readme and package.xml, as mentioned I'd like to keep these links for my repo. If you change this after you have merged the pull request, that would be great. In the code itself you don't have to change the address. The code looks in the .git directory for the correct address.

You might need to copy my wiki content to yours. I don't believe this will be done automatically by the pull request. 
The helpbutton in the Ribbon Design menu is looking for a wiki page. If you don't want use a wiki page you have to change:  
- line 414 in FCBinding.py
- line 1628 in LoadDesign_Ribbon.py
  
kind regards,

Paul

**Comment by geolta** [comment]
Nice, I will merge this now :)
